### ISSUE TITLE
[rush] dependencies missing after rush deploy

### ISSUE THREAD

**Issue by willsoto** [microsoft/rushstack#1943]
**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

Applications that run under normal circumstances fail to run after `rush deploy`. This occurs under `pnpm` `v4.x` and `v5.x`

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

Minimal reproduction: https://github.com/willsoto/rush-dep-repro

Only changes from default settings are the [pnpmVersion](https://github.com/willsoto/rush-dep-repro/blob/main/rush.json#L29), [rush version](https://github.com/willsoto/rush-dep-repro/blob/main/rush.json#L19) and I enable [strictPeerDependencies](https://github.com/willsoto/rush-dep-repro/blob/main/rush.json#L64)

Steps to reproduce:

1. `rush update`
2. `rush build`
3. `rush deploy`
4. `cd common/deploy/apps/api`
5. `rushx start:prod`

Output:

```
❯ pwd
/code/rush-dep-repro/common/deploy/apps/api
```

```
Found configuration in /code/rush-dep-repro/rush.json


Rush Multi-Project Build Tool 5.26.0 - https://rushjs.io
Node.js version is 12.18.0 (LTS)

Found configuration in /code/rush-dep-repro/rush.json

Executing: "node dist/main"

internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '@nestjs/common'
Require stack:
- /code/rush-dep-repro/common/deploy/common/temp/node_modules/.pnpm/nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1/node_modules/nest-raven/dist/raven.interceptor.js
- /code/rush-dep-repro/common/deploy/common/temp/node_modules/.pnpm/nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1/node_modules/nest-raven/dist/index.js
- /code/rush-dep-repro/common/deploy/apps/api/dist/app.module.js
- /code/rush-dep-repro/common/deploy/apps/api/dist/main.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Module.require (internal/modules/cjs/loader.js:1026:19)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (/code/rush-dep-repro/common/deploy/common/temp/node_modules/.pnpm/nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1/node_modules/nest-raven/dist/raven.interceptor.js:32:18)
    at Module._compile (internal/modules/cjs/loader.js:1138:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1158:10)
    at Module.load (internal/modules/cjs/loader.js:986:32)
    at Function.Module._load (internal/modules/cjs/loader.js:879:14)
    at Module.require (internal/modules/cjs/loader.js:1026:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/code/rush-dep-repro/common/deploy/common/temp/node_modules/.pnpm/nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1/node_modules/nest-raven/dist/raven.interceptor.js',
    '/code/rush-dep-repro/common/deploy/common/temp/node_modules/.pnpm/nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1/node_modules/nest-raven/dist/index.js',
    '/code/rush-dep-repro/common/deploy/apps/api/dist/app.module.js',
    '/code/rush-dep-repro/common/deploy/apps/api/dist/main.js'
  ]
}
```

**What is the expected behavior?**

This is the output when running from `apps/api`:

```
❯ rushx start:prod
Found configuration in /code/rush-dep-repro/rush.json


Rush Multi-Project Build Tool 5.26.0 - https://rushjs.io
Node.js version is 12.18.0 (LTS)

Found configuration in /code/rush-dep-repro/rush.json

Executing: "node dist/main"

[Nest] 97678   - 06/18/2020, 7:35:57 PM   [NestFactory] Starting Nest application...
[Nest] 97678   - 06/18/2020, 7:35:57 PM   [InstanceLoader] RavenModule dependencies initialized +11ms
[Nest] 97678   - 06/18/2020, 7:35:57 PM   [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 97678   - 06/18/2020, 7:35:57 PM   [RoutesResolver] AppController {}: +5ms
[Nest] 97678   - 06/18/2020, 7:35:57 PM   [RouterExplorer] Mapped {, GET} route +2ms
[Nest] 97678   - 06/18/2020, 7:35:57 PM   [NestApplication] Nest application successfully started +2ms
```

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** pnpm
* **Tool Version:** v4.14.4 and v5.2.1
* **Node Version:** v12.18.1 and v14.3.0
  * **Is this a LTS version?** v12 is an LTS version
  * **Have you tested on a LTS version?** yes
* **OS:** MacOS 10.15.3

**Comment by octogonz** [comment]
@willsoto This is a pretty vanilla phantom dependency issue.

The `nest-raven` package loads `@nestjs/common` from this file:

**nest-raven\dist\raven.interceptor.js**
```js
Object.defineProperty(exports, "__esModule", { value: true });
exports.RavenInterceptor = void 0;
const common_1 = require("@nestjs/common");  // <======
const operators_1 = require("rxjs/operators");
const Sentry = __importStar(require("@sentry/minimal"));
const node_1 = require("@sentry/node");
const graphql_1 = require("@nestjs/graphql");
let RavenInterceptor = (() => {
    let RavenInterceptor = class RavenInterceptor {
```

Whereas the `nest-raven` package declares `@nestjs/common` as a dev dependency:

**common\temp\node_modules\.pnpm\nest-raven@6.0.0-alpha.2_@nestjs+graphql@7.4.1\node_modules\nest-raven\package.json**
```js
{
  "name": "nest-raven",
  "version": "6.0.0-alpha.2",
  . . .
  "dependencies": {
    "@sentry/minimal": "^5.12.0",
    "@sentry/node": "^5.12.3"
  },
  "peerDependencies": {
    "@nestjs/graphql": "^6.2.4 || ^7.0.0"
  },
  "devDependencies": {
    "@nestjs/common": "^7.0.8",  // <====== incorrect
    "@nestjs/core": "^7.0.8",
  . . .
```

If it's needed at runtime, then it should have been declared as a regular dependency.

It works in the main monorepo because PNPM creates a workaround symlink `common\temp\node_modules\@nestjs\common` that enables it to be found even though it wasn't declared.  But `rush deploy` doesn't consider that.

I'm guessing that fixing https://github.com/microsoft/rushstack/issues/1942 will handle your case here.  

I'm not going to resolve it as a dupe however, since we can use your repro to test that fix.

**Comment by octogonz** [comment]
This is a great repro. 😋

After fixing `@nestjs\common`, I encountered another case where `@nestjs\graphql` requires `apollo-graphql` in this file:

**@nestjs\graphql\dist\federation\graphql-federation.factory.js**
```js
. . .
const shared_utils_1 = require("@nestjs/common/utils/shared.utils");
const apollo_graphql_1 = require("apollo-graphql");  // <==========
const apollo_server_core_1 = require("apollo-server-core");
const graphql_1 = require("graphql");
const graphql_tools_1 = require("graphql-tools");
. . .
```

...but it's not declared anywhere in the **package.json** file, not even as a dev dependency:

**common\temp\node_modules\.pnpm\@nestjs\graphql@7.4.1_d9776d8df05615f886c33564b99f3d72\node_modules\@nestjs\graphql\package.json**

AND in this case there is no PNPM workaround link in **common/temp/node_modules**; instead the workaround link is located under **common/temp/node_modules/.pnpm**.

**Comment by octogonz** [comment]
Here's a prototype of a fix for this issue: PR https://github.com/microsoft/rushstack/pull/1949

With this fix, your `rushx start:prod` command correctly boots up the Nest application successfully. 🥳

**Comment by willsoto** [comment]
That is great! Glad it helped. 

Does it make sense to file issues upstream for these two repos? Seems like these dependencies should be declared somewhere on their end...

**Comment by octogonz** [comment]
> Does it make sense to file issues upstream for these two repos? Seems like these dependencies should be declared somewhere on their end...

Sure. It's sometimes not easy to get people to understand that they are doing something wrong, though.

**Comment by octogonz** [comment]
This fix was released with Rush 5.27.2
